src/plugins/discover/public/application/main/components/layout/discover_documents.tsx

```diff
@@ -75,49 +75,75 @@ function DiscoverDocumentsComponent({
   onAddFilter?: DocViewFilterFn;
   savedSearch: SavedSearch;
   setExpandedDoc: (doc?: DataTableRecord) => void;
-  state: AppState;
-  stateContainer: GetStateReturn;
+  stateContainer: DiscoverStateContainer;
   onFieldEdited?: () => void;
 }) {
   const { capabilities, dataViews, uiSettings } = useDiscoverServices();
+  const [query, sort, rowHeight, rowsPerPage, grid, columns, index] = useAppStateSelector(
+    (state) => {
+      return [
+        state.query,
+        state.sort,
+        state.rowHeight,
+        state.rowsPerPage,
+        state.grid,
+        state.columns,
+        state.index,
+      ];
+    }
+  );
+
   const useNewFieldsApi = useMemo(() => !uiSettings.get(SEARCH_FIELDS_FROM_SOURCE), [uiSettings]);
   const hideAnnouncements = useMemo(() => uiSettings.get(HIDE_ANNOUNCEMENTS), [uiSettings]);
   const isLegacy = useMemo(() => uiSettings.get(DOC_TABLE_LEGACY), [uiSettings]);
   const sampleSize = useMemo(() => uiSettings.get(SAMPLE_SIZE_SETTING), [uiSettings]);
 
   const documentState = useDataState(documents$);
-  const isLoading = documentState.fetchStatus === FetchStatus.LOADING;
-  const isPlainRecord = useMemo(
-    () => getRawRecordType(state.query) === RecordRawType.PLAIN,
-    [state.query]
-  );
+  const isDataLoading = documentState.fetchStatus === FetchStatus.LOADING;
+  // This is needed to prevent EuiDataGrid pushing onSort because the data view has been switched.
+  // 1. When switching the data view, the sorting in the URL is reset to the default sorting of the selected data view.
+  // 2. The new sort param is already available in this component and propagated to the EuiDataGrid.
+  // 3. currentColumns are still referring to the old state
+  // 4. since the new sort by field isn't available in currentColumns EuiDataGrid is emitting a 'onSort', which is unsorting the grid
```

Comment:
Here's where EuiDataGrid calls the onSort event: 

https://github.com/elastic/eui/blob/dff7a774b7065119838a3791ae77c9536f2ccf4c/src/components/datagrid/controls/column_sorting.tsx#L41-L59